# Tipos Básicos de Variáveis

Python é uma linguagem de tipagem dinâmica - isso significa que você não
precisa especificar com antecedência que tipo de dado você vai armazenar
em uma variável.  No entanto, existem alguns tipos de dados principais
com os quais precisamos nos familiarizar ao usar a linguagem.

O primeiro conjunto de tipos de dados é semelhante àqueles encontrados em outras
linguagens (como C/C++ e Fortran): números de ponto flutuante, inteiros
e strings.

```{tip}
O ponto flutuante é essencial para a ciência computacional.  Uma ótima
introdução ao ponto flutuante e suas limitações é: [O que todo
cientista da computação deveria saber sobre aritmética de ponto flutuante](http://dl.acm.org/citation.cfm?id=103163) de
D. Goldberg. 
```

O próximo conjunto de tipos de dados são containers.  Em Python, ao contrário de algumas
linguagens, estes são integrados à linguagem e tornam muito fácil realizar
operações complexas.  Vamos ver isso mais adiante.

Alguns exemplos vêm do tutorial oficial do Python:
http://docs.python.org/3/tutorial/

## Números Inteiros

Números inteiros são números sem um ponto decimal. Eles podem ser positivos ou negativos. A maioria das linguagens de programação usa uma quantidade finita de memória para armazenar um único inteiro, mas em Python, a quantidade de memória será expandida conforme necessário para armazenar inteiros grandes.

Os operadores básicos, `+`, `-`, `*` e `/` funcionam com inteiros.

In [ ]:
2+2+3

7

In [ ]:
2*-4

-8

```{note}
A divisão inteira é um ponto onde o Python 2 e o Python 3 são diferentes

No Python 3.x, dividir 2 inteiros resulta em um número de ponto flutuante.  No Python 2.x, dividir 2 inteiros resulta em um número inteiro.  O último é consistente com muitas linguagens de programação fortemente tipadas (como Fortran ou C), já que o tipo de dado do resultado é o mesmo das entradas, mas o primeiro está mais alinhado com as nossas expectativas.
```

In [ ]:
3/2

1.5

Para obter um resultado inteiro, usamos o operador //

In [ ]:
3//2

1

Python é uma _linguagem de tipagem dinâmica_—isso significa que não precisamos declarar o tipo de dado de uma variável antes de inicializá-la.  

Aqui vamos criar uma variável (pense nela como um rótulo descritivo que pode se referir a algum dado). O operador `=` atribui um valor à variável.

In [ ]:
a = 1.5
b = 3

Funções operam sobre variáveis e retornam um resultado. Aqui, `print()` irá exibir a saída na tela.

In [ ]:
a + b

4.5

In [ ]:
a * b

4.5

Observe que os nomes de variáveis diferenciam maiúsculas de minúsculas, então `a` e `A` são diferentes.

In [ ]:
A = 2025

In [ ]:
print(a, A)

1.5 2025


Aqui inicializamos 3 variáveis, todas com o valor `0`, mas essas ainda são variáveis distintas, então podemos mudar uma sem afetar as outras.

In [ ]:
x = y = z = 0

In [ ]:
print(x, y, z)

0 0 0


In [ ]:
z = 1

In [ ]:
z

1

O Python tem alguma ajuda embutida (e o Jupyter/IPython tem ainda mais).

Tente fazer:
```
help(x)
```

In [ ]:
help(z)

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |

Outra função, `type()`, retorna o tipo de dado de uma variável

In [ ]:
type(x)

int

```{note}
Observe que em linguagens como Fortran e C, você especifica a quantidade de memória que um inteiro pode ocupar (geralmente 2 ou 4 bytes). Isso impõe uma restrição sobre o maior tamanho de inteiro que pode ser representado. O Python adaptará o tamanho do inteiro para que você não tenha *overflow*.
```

In [ ]:
a = 12345678901234567890123456789012345123456789012345678901234567890
print(a)
print(a.bit_length())
print(type(a))

12345678901234567890123456789012345123456789012345678901234567890
213
<class 'int'>


## Ponto Flutuante

Ao operar com números de ponto flutuante e inteiros, o resultado é promovido a um float.

In [ ]:
1. + 2

3.0

Mas observe o operador especial de divisão inteira.

In [ ]:
1.//2

0.0

```{important}
É importante entender que, como existem infinitos números reais entre quaisquer dois limites, em um computador precisamos aproximar isso por um número finito. Existe um padrão IEEE para ponto flutuante que praticamente todas as linguagens e processadores seguem.

Isso significa duas coisas:

* Nem todo número real terá uma representação exata em ponto flutuante.
* Há uma precisão finita para os números — abaixo disso, perdemos o controle das diferenças (isso é geralmente chamado de *erro de arredondamento*).
```

Este artigo é uma referência incrível para entender como um computador armazena números:

(https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)

Considere a seguinte expressão, por exemplo:

In [ ]:
0.3/0.1 - 3

-4.440892098500626e-16

In [ ]:
0.3/0.1 == 3

False

Aqui está outro exemplo: O número `0.1` não pode ser representado exatamente em um computador. Em nossa impressão, usamos um especificador de formato (as coisas dentro de `{}`) para pedir que mais precisão seja mostrada:

In [ ]:
a = 0.1
print("{:30.20}".format(a))

        0.10000000000000000555


Podemos pedir ao Python para relatar os limites do ponto flutuante.

In [ ]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

Observe que isso diz que só podemos armazenar números entre `2.2250738585072014e-308` e `1.7976931348623157e+308`.

Também vemos que a precisão é `2.220446049250313e-16` (isso é comumente chamado de _epsilon da máquina_). Para ver isso, considere adicionar um número pequeno a `1.0`. Usaremos o operador de igualdade (`==`) para testar se dois números são iguais:

```{admonition} Exercício Rapído

Defina duas variáveis, \( a = 1 \) e \( e = 10^{-16} \).

Agora defina uma terceira variável, `b = a + e`.

Podemos usar o operador `==` do Python para testar a igualdade. O que você espera que `b == a` retorne? Execute e veja se concorda com seu palpite.
```

## Módulos e Namespaces

A linguagem Python básica é bastante pequena. A maior parte acontece em módulos. Alguns módulos fazem parte de uma biblioteca padrão que fornece funcionalidade adicional. Essas partes adicionadas estão na forma de módulos que podemos _importar_ para nossa sessão (ou programa) Python.

O módulo `math` fornece funções que realizam as operações matemáticas básicas, bem como constantes (note que há um módulo separado `cmath` para números complexos).

Em Python, você `importa` um módulo. As funções são então definidas em um _namespace_ separado — esta é uma região separada que define nomes e variáveis, etc. Uma variável em um namespace pode ter o mesmo nome que uma variável em um namespace diferente, e elas não entram em conflito. Você usa o operador `.` para acessar um membro de um namespace.

Por padrão, quando você digita algo no interpretador Python ou aqui no notebook Jupyter, ou em um script, ele está em seu próprio namespace padrão, e você não precisa prefixar nenhuma das variáveis com um indicador de namespace.

In [ ]:
import math

O módulo `math` fornece o valor de π (pi).

In [ ]:
math.pi

3.141592653589793

Isso é distinto de qualquer variável `pi` que possamos definir aqui.

In [ ]:
pi = 3

In [ ]:
print(pi, math.pi)

3 3.141592653589793


Observe que `pi` e `math.pi` são distintos um do outro—eles estão em namespaces diferentes.

### Operações de Ponto Flutuante

Os mesmos operadores, `+`, `-`, `*`, `/` funcionam como de costume para números de ponto flutuante. Para elevar um número a uma potência, usamos o operador `**` (isso é o mesmo que em Fortran).

In [ ]:
R = 2.0

In [ ]:
math.pi * R**2

12.566370614359172

A precedência dos operadores segue a de muitas linguagens. Veja [Precedência de Operadores do Python](https://docs.python.org/3/reference/expressions.html#operator-precedence) para mais detalhes.

Em ordem de precedência:
* Quantidades em `()`
* Fatiamento, chamadas, subscritos
* Exponenciação (`**`)
* `+x`, `-x`, `~x`
* `*`, `@`, `/`, `//`, `%`
* `+`, `-`

(Depois disso estão as operações bitwise e comparações)

Os parênteses podem ser usados para sobrepor a precedência.

<div class="alert alert-block alert-warning">
    
<span class="fa fa-flash"></span> Quick Exercise:

Consider the following expressions.  Using the ideas of precedence, think about what value will result, then try it out in the cell below to see if you were right.

  * `1 + 3*2**2`
  * `1 + (3*2)**2`
  * `2**3**2`

</div>

```{admonition} Exercício Rapído
Considere as seguintes expressões. Usando as ideias de precedência, pense sobre qual valor resultará e, em seguida, experimente no código abaixo para ver se você estava certo.

* `1 + 3*2**2`
* `1 + (3*2)**2`
* `2**3**2`
```

O módulo `math` fornece muitas das funções matemáticas padrão. A maioria delas é na verdade repetida no `numpy`, então, na prática, eu pessoalmente quase nunca uso `math`.

Para as funções trigonométricas, a expectativa é que o argumento da função esteja em radianos—você pode usar `math.radians()` para converter de graus para radianos, por exemplo:

In [ ]:
math.cos(math.radians(45))

0.7071067811865476

Observe que nessa declaração estamos alimentando a saída de uma função (`math.radians()`) em uma segunda função, `math.cos()`

Quando em dúvida, peça ajuda para descobrir todas as coisas que um módulo fornece:

In [ ]:
help(math.sin)

Help on built-in function sin in module math:

sin(x, /)
    Return the sine of x (measured in radians).



## Números Complexos

O Python usa '`j`' para denotar a unidade imaginária.

In [ ]:
1.0 + 2j

(1+2j)

In [ ]:
a = 1j
b = 3.0 + 2.0j
print(a + b)
print(a * b)

(3+3j)
(-2+3j)


Podemos usar `abs()` para obter a magnitude e, separadamente, obter as partes real ou imaginária.

In [ ]:
print("magnitude: ", abs(b))
print("real part: ", a.real)
print("imag part: ", a.imag)

magnitude:  2
real part:  0.1
imag part:  0.0


## Strings

O Python não se importa se você usa aspas simples ou duplas para strings:

In [ ]:
a = "está é a minha string"
b = 'outra string'

In [ ]:
print(a)
print(b)

está é a minha string
outra string


Muitos dos operadores matemáticos usuais também são definidos para strings. Por exemplo, para concatenar ou duplicar:

In [ ]:
a + b

'está é a minha stringoutra string'

In [ ]:
a + ". " + b

'está é a minha string. outra string'

In [ ]:
a * 3

'está é a minha stringestá é a minha stringestá é a minha string'

Existem vários códigos de escape que são interpretados em strings. Esses começam com uma barra invertida, `\`. Por exemplo, você pode usar `\n` para uma nova linha.

In [ ]:
a = a + "\n" + "hello"
print(a)

está é a minha string
hello


```{admonition} Exercício Rapído
    
A função `input()` pode ser usada para pedir a entrada do usuário.

* Use `help(input)` para ver como funciona.  
* Escreva um código para pedir a entrada e armazenar o resultado em uma variável. `input()` retornará uma string.

* Use a função `float()` para converter um número inserido como entrada em uma variável de ponto flutuante.  
* Verifique se a conversão funcionou usando a função `type()`.
```

Aspas triplas `"""` podem envolver strings de várias linhas. Isso é útil para docstrings no início das funções (mais sobre isso mais tarde...).

In [ ]:
c = """
É um período de guerra civil. Naves rebeldes, 
atacando de uma base oculta, conquistaram sua primeira 
vitória contra o maligno Império Galáctico.

Durante a batalha, espiões rebeldes conseguiram roubar 
os planos secretos da arma definitiva do Império, a 
Estrela da Morte, uma estação espacial blindada com poder 
suficiente para destruir um planeta inteiro.

Perseguida pelos sinistros agentes do Império, a Princesa 
Leia corre para casa a bordo de sua nave estelar, guardiã 
dos planos roubados que podem salvar seu povo e 
restaurar a liberdade na galáxia..."""

In [ ]:
print(c)


É um período de guerra civil. Naves rebeldes, 
atacando de uma base oculta, conquistaram sua primeira 
vitória contra o maligno Império Galáctico.

Durante a batalha, espiões rebeldes conseguiram roubar 
os planos secretos da arma definitiva do Império, a 
Estrela da Morte, uma estação espacial blindada com poder 
suficiente para destruir um planeta inteiro.

Perseguida pelos sinistros agentes do Império, a Princesa 
Leia corre para casa a bordo de sua nave estelar, guardiã 
dos planos roubados que podem salvar seu povo e 
restaurar a liberdade na galáxia...


Uma string bruta não substitui sequências de escape (como `\n`). Basta colocar um `r` antes da primeira aspa:

In [ ]:
d = r"está é uma string bruta \n hello"
d

'está é uma string bruta \\n hello'

O fatiamento é usado para acessar uma parte de uma string.

Fatiar uma string pode parecer um pouco contra-intuitivo se você vem de C ou Fortran. O truque é pensar no índice como representando a borda esquerda de um caractere na string. Quando fizermos arrays mais tarde, o mesmo se aplicará.

Observe também que o Python (como C) usa indexação baseada em 0.

Os índices negativos contam a partir da direita.

In [ ]:
# Definindo uma string de exemplo
a = "está é a minha string"

# Imprimindo a string completa
print(a) 

# Imprimindo uma parte da string usando fatiamento
# a[11:14] significa: comece no índice 11 e vá até o índice 14 (exclusivo)
print(a[11:14]) 

# Imprimindo o primeiro caractere da string
# a[0] significa: pegue o caractere no índice 0
print(a[0])


# Imprimindo o segundo caractere a partir do final da string 'a'
print(a[-2])

# Imprimindo uma parte da string usando fatiamento com passo
# a[11:20:2] significa: comece no índice 11, vá até o índice 20 (exclusivo), pegando a cada 2 caracteres
print(a[11:20:2])  # Saída: "mni" (caracteres nos índices 11, 13, 15, 17, 19)

está é a minha string
nha
e
n
nasrn


```{admonition} Quick Exercise

As strings têm muitos _métodos_ (funções que sabem como trabalhar com um tipo de dado específico, neste caso, strings). Um método útil é `.find()`. Para uma string `a`, `a.find(s)` retornará o índice da primeira ocorrência de `s`.

Para nossa string `c` acima, encontre o primeiro `.` (identificando a primeira frase completa) e imprima apenas a primeira frase em `c` usando esse resultado.

```

Existem também vários métodos e funções que trabalham com strings. Aqui estão alguns exemplos:

In [ ]:
print(a)  # Exibe a string original

print(a.replace("está", "aquela"))  # Exibe a string com "esta" substituído por "aquela"

print(len(a))  # Exibe o número total de caracteres na string

print(a)  # Exibe a string original novamente, sem alterações

está é a minha string
aquela é a minha string
21
está é a minha string


Note que nossa string original, `a`, não foi alterada. Em Python, as strings são *imutáveis*. Operações em strings retornam uma nova string.

In [ ]:
type(a)

str

Podemos formatar strings ao imprimir para inserir quantidades em lugares específicos na string. Um `{}` serve como um espaço reservado para uma quantidade e é substituído usando o método `.format()`:

In [ ]:
a = 1
b = 2.0
c = "test"
print("a = {}; b = {}; c = {}".format(a, b, c))

a = 1; b = 2.0; c = test


Poderiamos juntar string também usando `+`

In [ ]:
print("a="+str(a))

a=1


Mas a maneira mais moderna de fazer isso é usar *f-strings*.

In [ ]:
print(f"a = {a}; b = {b}; c = {c}")

a = 1; b = 2.0; c = test


Note o `f` precedendo a aspa inicial `"` 

Uma outra opção é a seguinte, onde você pode especificar o formato de cada variável:

In [ ]:
print("a = %i; b = %.2f; c = %s" % (a, b, c))

a = 1; b = 2.00; c = test
